File: htdocs/database/BigliettoDAO.php
<?php

require_once __DIR__ . '/DBManager.php';
require_once __DIR__ . '/../entity/EntityBiglietto.php';

class BigliettoDAO {
   /**
   * Recupera il biglietto avente un certo id.
   * @param id id del biglietto
   * @return EntityBiglietto|null
   * @throws Exception
   */
    public static function readById(int $id): ?EntityBiglietto {
        if($id <= 0)
            throw new Exception('ID Biglietto non valido');

        try{
            $db = DBManager::getInstance()->getConnection();

            $stmt = $db->prepare(
                'SELECT * FROM biglietti WHERE ID = :id'
            );
            $stmt->execute(['id' => $id]);
            $row = $stmt->fetch();
            if (!$row) return null;
            return new EntityBiglietto(
                (int)$row['ID'], // ID biglietto
                (int)$row['Corsa'], // ID Corsa
                (float)$row['Prezzo'], // Prezzo
                new DateTime($row['Data']), // Data
                $row['Cliente'], // Nome utente
                (int)$row['Impiegato']  // ID Impiegato
            );            
        }catch(Exception $e){
            throw new Exception('Errore durante la lettura del biglietto.');
        }

    }

   /**
   * Recupera i biglietti acquistati da un cliente.
   * @param nomeUtente username del cliente
   * @return EntityBiglietto[]
   * @throws Exception
   */
    public static function readByCliente(String $nomeUtente): array {
        if($nomeUtente === null)
            throw new Exception('nome utente non valido');

        try{
            $db = DBManager::getInstance()->getConnection();

            $stmt = $db->prepare(
                'SELECT * FROM biglietti WHERE Cliente = :cid'
            );
            $stmt->execute(['cid' => $nomeUtente]);
            $list = [];
            while ($row = $stmt->fetch()) {
                $list[] = new EntityBiglietto(
                    (int)$row['ID'], // ID Biglietto
                    (int)$row['Corsa'], // ID Corsa
                    (float)$row['Prezzo'], // Prezzo di emissione del biglietto
                    new DateTime($row['Data']), // Data di emissione del biglietto
                    (int)$row['Impiegato'], // ID Impiegato
                    (int)$row['Cliente'] // Nome Utente Cliente
                );
            }
            return $list;            
        }catch(Exception $e){
            throw new Exception('Errore durante la lettura del biglietto.');
        }

    }

   /**
   * Salva un biglietto.
   * @param biglietto biglietto da salvare
   * @return EntityBiglietto
   * @throws Exception
   */
    public static function createBiglietto(EntityBiglietto $biglietto): EntityBiglietto {
        if($biglietto === null)
            throw new Exception('biglietto non valido');

        try{
            $db = DBManager::getInstance()->getConnection();

            $stmt = $db->prepare(
                'INSERT INTO biglietti (Corsa, Prezzo, Data, Impiegato, Cliente)
                VALUES (:corsa, :prezzo, :data, :impiegato, :cliente)'
            );

            $stmt->execute([
                'corsa'    => $biglietto->getIdCorsa(),
                'prezzo'   => $biglietto->getPrezzoVendita(),
                'data'     => $biglietto->getDataEmissione()->format('Y-m-d H:i:s'),
                'impiegato'=> $biglietto->getIdImpiegato(),
                'cliente'  => $biglietto->getIdClienteRegistrato()
            ]);
            $biglietto->setIdBiglietto((int)$db->lastInsertId());

            return $biglietto;            
        }catch(Exception $e){
            throw new Exception('Errore durante il salvataggio del biglietto.');
        }

    }

   /**
   * Restituisce il numero di biglietti emessi per una data corsa.
   * @param id id della corsa
   * @return int|null
   * @throws Exception
   */
    public static function bigliettiPerCorsa(int $id): ?int {
        if($id <= 0)
            throw new Exception('ID Corsa non valido');
        
        try{
            $db = DBManager::getInstance()->getConnection();

            $stmt = $db->prepare(
                'SELECT COUNT(ID) AS tot FROM biglietti WHERE Corsa = :id'
            );

            $stmt->execute(['id' => $id]);
            $row = $stmt->fetch();
            if (!$row) return null;

            return $row['tot'];            
        }catch(Exception $e){
            throw new Exception('Errore durante la lettura del biglietto.');
        }

    }
}
?>

File: src/it/nominasuntsubstantiarerum/netbus/database/BigliettoDAO.java
package it.nominasuntsubstantiarerum.netbus.database;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;

import it.nominasuntsubstantiarerum.netbus.entity.EntityBiglietto;
import it.nominasuntsubstantiarerum.netbus.exception.DAOException;
import it.nominasuntsubstantiarerum.netbus.exception.DBConnectionException;

public class BigliettoDAO {
	public static void createBiglietto(EntityBiglietto biglietto) throws DAOException, DBConnectionException {
		if(biglietto == null)
			throw new IllegalArgumentException("biglietto non valido");
		
		try {
			Connection connection = DBManager.getConnection();
			String query = "INSERT INTO biglietti (corsa, prezzo, data, impiegato) VALUES (?, ?, ?, ?)";
			
			try {
				PreparedStatement preparedStatement = connection.prepareStatement(query);
				preparedStatement.setInt(1, biglietto.getIdCorsa());
				preparedStatement.setFloat(2, biglietto.getPrezzoVendita());
				preparedStatement.setDate(3, new java.sql.Date(biglietto.getDataEmissione().getTime()));
				preparedStatement.setString(4, biglietto.getIdImpiegato());
				preparedStatement.executeUpdate();
			} catch (SQLException e) {
				throw new DAOException("Errore durante l'inserimento del biglietto.");
			} finally {
				DBManager.closeConnection();
			}
		} catch (SQLException e) {
			throw new DBConnectionException("Errore di connessione al database.");
		}
	}

	public static EntityBiglietto readBiglietto(int id) throws DAOException, DBConnectionException {
		if(id <= 0)
			  throw new IllegalArgumentException("ID biglietto non valido: " + id);
		
		try {
			Connection connection = DBManager.getConnection();
			String query = "SELECT * FROM biglietti WHERE id = ?";
			
			try {
				PreparedStatement preparedStatement = connection.prepareStatement(query);
				preparedStatement.setInt(1, id);
				ResultSet resultSet = preparedStatement.executeQuery();
				
				if (resultSet.next()) {
					int bId = resultSet.getInt("id");
					int corsa = resultSet.getInt("corsa");
					float prezzo = resultSet.getFloat("prezzo");
					java.sql.Date data = resultSet.getDate("data");
					String impiegato = resultSet.getString("impiegato");
					String idCliente = resultSet.getString("cliente");
					return new EntityBiglietto(bId, corsa, prezzo, new java.util.Date(data.getTime()), impiegato, idCliente);
				} else {
					throw new DAOException("Biglietto non trovato.");
				}
			} catch (SQLException e) {
				throw new DAOException("Errore durante la lettura del biglietto.");
			} finally {
				DBManager.closeConnection();
			}
		} catch (SQLException e) {
			throw new DBConnectionException("Errore di connessione al database.");
		}
	}
	
	/**
	 * Metodo per creare più biglietti in un'unica operazione.
	 * @param biglietto Biglietto da emettere
	 * @param ripetizioni Numero di biglietti da emettere
	 * @throws DAOException Se si verifica un errore durante l'inserimento dei biglietti
	 * @throws DBConnectionException Se si verifica un errore di connessione al database
	 * @return L'ID dell'ultimo biglietto inserito
	 */
	public static int createBiglietti(EntityBiglietto biglietto, int ripetizioni) throws DAOException, DBConnectionException {
		try {
			Connection connection = DBManager.getConnection();
			String query = "INSERT INTO biglietti (corsa, prezzo, data, impiegato) VALUES (?, ?, ?, ?)";
			String returnQuery = "SELECT LAST_INSERT_ID()"; // Per ottenere l'ID dell'ultimo biglietto inserito
			try {
				PreparedStatement preparedStatement = connection.prepareStatement(query);
				preparedStatement.setInt(1, biglietto.getIdCorsa());
				preparedStatement.setFloat(2, biglietto.getPrezzoVendita());
				preparedStatement.setDate(3, new java.sql.Date(biglietto.getDataEmissione().getTime()));
				preparedStatement.setString(4, biglietto.getIdImpiegato());
				
				for (int i = 0; i < ripetizioni; i++) 
					preparedStatement.executeUpdate();
				
				PreparedStatement returnStatement = connection.prepareStatement(returnQuery);
				ResultSet resultSet = returnStatement.executeQuery();
				if (resultSet.next()) {
					return resultSet.getInt(1);  // Restituisce l'ID dell'ultimo biglietto inserito
				} else {
					throw new DAOException("Nessun biglietto inserito.");
				}
			} catch (SQLException e) {
				throw new DAOException("Errore durante l'inserimento dei biglietti.");
			} finally {
				DBManager.closeConnection();
			}
		} catch (SQLException e) {
			throw new DBConnectionException("Errore di connessione al database.");
		}
	}
	
	public static int bigliettiPerCorsa(int idCorsa) throws DAOException, DBConnectionException {
		if(idCorsa <= 0)
			throw new IllegalArgumentException("ID corsa non valido: " + idCorsa);
		
		try {
			Connection connection = DBManager.getConnection();
			String query = "SELECT COUNT(1) AS numero FROM biglietti WHERE corsa = ?";
			
			try {
				PreparedStatement preparedStatement = connection.prepareStatement(query);
				preparedStatement.setInt(1, idCorsa);
				ResultSet resultSet = preparedStatement.executeQuery();
				
				if (resultSet.next()) {
					return resultSet.getInt("numero");
				} else {
					throw new DAOException("Nessun biglietto trovato per la corsa con ID " + idCorsa);
				}
			} catch (SQLException e) {
				throw new DAOException("Errore durante il conteggio dei biglietti per la corsa.");
			} finally {
				DBManager.closeConnection();
			}
		} catch (SQLException e) {
			throw new DBConnectionException("Errore di connessione al database.");
		}
	}
}
